crates/gpui/src/platform/windows/dispatcher.rs

```diff
@@ -1,74 +1,119 @@
-        let threadpool = unsafe {
-            let ret = CreateThreadpool(None);
-            if ret.0 == 0 {
-                panic!(
-                    "unable to initialize a thread pool: {}",
-                    std::io::Error::last_os_error()
-                );
-            }
-            // set minimum 1 thread in threadpool
-            let _ = SetThreadpoolThreadMinimum(ret, 1)
-                .inspect_err(|_| log::error!("unable to configure thread pool"));
-
-            ret
+    pub(crate) fn new() -> Self {
+        // Windows 10 Fall Creators Update (introduced in 10.0.16299.0)
+        let controller = unsafe {
+            let options = DispatcherQueueOptions {
+                dwSize: std::mem::size_of::<DispatcherQueueOptions>() as u32,
+                threadType: DQTYPE_THREAD_CURRENT,
+                apartmentType: DQTAT_COM_NONE,
+            };
+            CreateDispatcherQueueController(options).unwrap()
         };
+        let main_queue = controller.DispatcherQueue().unwrap();
+        let parker = Mutex::new(Parker::new());
         let main_thread_id = current().id();
+
         WindowsDispatcher {
-            threadpool,
-            main_sender,
+            controller,
+            main_queue,
             parker,
             main_thread_id,
-            dispatch_event,
         }
     }
 
     fn dispatch_on_threadpool(&self, runnable: Runnable) {
-        unsafe {
-            let ptr = Box::into_raw(Box::new(runnable));
-            let environment = get_threadpool_environment(self.threadpool);
-            let Ok(work) =
-                CreateThreadpoolWork(Some(threadpool_runner), Some(ptr as _), Some(&environment))
-                    .inspect_err(|_| {
-                        log::error!(
-                            "unable to dispatch work on thread pool: {}",
-                            std::io::Error::last_os_error()
-                        )
-                    })
-            else {
-                return;
-            };
-            SubmitThreadpoolWork(work);
-        }
+        let runner = generate_handler!(WorkItemHandler, runnable, true);
+        ThreadPool::RunWithPriorityAndOptionsAsync(
+            &runner,
+            WorkItemPriority::High,
+            WorkItemOptions::TimeSliced,
+        )
+        .log_err();
+    }
+
+    fn dispatch_on_threadpool_after(&self, runnable: Runnable, duration: Duration) {
+        let runner = generate_handler!(TimerElapsedHandler, runnable, true);
+        let delay = TimeSpan {
+            // A time period expressed in 100-nanosecond units.
+            // 10,000,000 ticks per second
+            Duration: (duration.as_nanos() / 100) as i64,
+        };
+        ThreadPoolTimer::CreateTimer(&runner, delay).log_err();
+    }
+}
+
+impl Drop for WindowsDispatcher {
+    fn drop(&mut self) {
+        self.controller.ShutdownQueueAsync().log_err();
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+struct TaskWrapper(*mut c_void);
```

Comment:
Do we need this wrapper struct? It seems like using a normal `Runnable` does the same job

Reply:
The compiler complains that `Runnable` is not `Send` nor `Sync`, so I wrapper it then unsafe impl `Send` for it.

Reply:
I'm unsure what's going on here then, as it seems like [it should implement Send and Sync for Runnable<()>](https://github.com/smol-rs/async-task/blob/master/src/runnable.rs#L694-L703) 🤔

We should have a better understanding of why these are different and what's going on before I can merge this.

Reply:
Sorry for the misinformation, the error is:
> cannot move out of `runnable`, a captured variable in an `FnMut` closure
move occurs because `runnable` has type `Runnable`, which does not implement the `Copy` trait

Here we are converting `Runnable` into a raw pointer, and then converting it back to `Runnable` inside the function, avoiding the need for a move operation. The `unsafe impl Send` here is for the `*mut c_void`.

```rust
    DispatcherQueueHandler::new(move || {
        runnable.run();     <- captured here
        Ok(())
    })
```

Reply:
Ok, if that's the case then I think we should inline the macro `generate_handler` code in all three places. Essentially, removing the macro and writing the code by hand, so I can evaluate what's going on with this move. 

The thing about avoiding the move operation with this technique is that you're just swapping in a copy operation instead. This means that somewhere in memory there are two pointers to a Runnable lurking, which could cause  undefined behavior.

Reply:
I have removed the macros. It's passing a pointer, I guess it's ok, this servers the same purpose as the macOS implement.

The macOS impl:
```rust
dispatch_async_f(
    dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH.try_into().unwrap(), 0),
    runnable.into_raw().as_ptr() as *mut c_void,
    Some(trampoline),
);

extern "C" fn trampoline(runnable: *mut c_void) {
    let task = unsafe { Runnable::<()>::from_raw(NonNull::new_unchecked(runnable as *mut ())) };
    task.run();
}
```

Reply:
Yes, but those functions  are externally defined c functions. `TimerElapsedHandler` is [explicitly defined in safe rust to take a `FnMut()`](https://microsoft.github.io/windows-docs-rs/doc/windows/System/Threading/struct.TimerElapsedHandler.html#impl-TimerElapsedHandler). Reading the documentation, this seems like an incorrect translation of API to Rust. But between the two, I'm going to trust the auto-generated rust code just in case. 

For us, I think that means I'd prefer a solution where we keep the Runnable type, wrap it in an `Option`, and then use `Option::take()` to get an owned Runnable, that we can consume with `Runnable::run()`

Reply:
Sorry for the slow response, I'm in China, and it's lunchtime now. Great idea, I forgot about this trick with the `Option`. With this approach, the code becomes very simple. Thank you so much!

```rust
let handler = {
    let mut task_wrapper = Some(runnable);
    WorkItemHandler::new(move |_| {
        task_wrapper.take().unwrap().run();
        Ok(())
    })
};
```

Reply:
Happy to :D 

And no worries, hope lunch was good!